File: TreeKalk/nodo.cpp

#include "nodo.h"



nodo::nodo()
{
    padre=NULL;
    info=double();
}
nodo::nodo(const nodo& i):info(i.info),padre(i.padre)     // costruttore di copia di nodo
{
    for(list<nodo*>::const_iterator it=i.nodi.begin();it!=i.nodi.end();it++)
        nodi.push_back((*it)->copy());
}
nodo::nodo(int num)               // costruttore che inizializza il nodo con num figli
{
    padre=NULL;
    info=double();
    for(int i=0;i<num;i++)
        nodi.push_back(NULL);
}
nodo::nodo(double i,nodo* p,list<nodo*>& n):info(i),padre(p),nodi(n)  // costruttore con info, padre e figli
{}
nodo::nodo(double i,nodo* p):info(i),padre(p){}       // costruttore con padre e info

nodo::~nodo()                             // distruttore di nodo
{
    //cout << "~nodo() cancello : " << endl;
    if(nodi.size()>0)
    {
        list<nodo*>::const_iterator it=nodi.begin();
        for(int i = 0;it!=nodi.end();it++,i++)
            if(*it)
                delete *it;
    }
}
nodo* nodo::getPadre()const    // ritorna il padre
{
    return padre;
}
double nodo::getInfo()const            // ritorna l'info del nodo
{
    return info;
}
int nodo::numeroFigli()const                  // ritorna il numero di figli di un nodo
{
    return nodi.size();
}
nodo* nodo::getNodo(unsigned int i)const    // ritorna il nodo iesimo se esiste
{
    if(!(nodi.size()) || i>=nodi.size())
        throw std::string("Nodo non esistente");
    list<nodo*>::const_iterator it=nodi.begin();
    unsigned int j=0;
    while(j<i)
    {
        j++;
        it++;
    }
    return *it;
}
void nodo::setPadre(nodo* n)       // metodo per settare il padre del nodo
{
    padre=n;
}
void nodo::setInfo(double i)    // metodo per settare info
{
    info=i;
}
void nodo::setFiglio(unsigned int j,nodo* n)    // permette di settare il figlio iesimo
{
    if(j>=nodi.size())
        throw std::string("Stai cercando di settare un figlio che non esiste");
    list<nodo*>::iterator it=nodi.begin();
    unsigned int i=0;
    while (i<j)
    {
        i++;
        it++;
    }
    /*if(*it)
        delete *it;*/
    *it=n;
}
nodo* nodo::operator[](unsigned int i)const
{
    if(i>=nodi.size())
        throw std::string("nodo non esistente");
    list<nodo*>::const_iterator it=nodi.begin();
    unsigned int j=0;
    while(j<i)
    {
        j++;
        it++;
    }
    return *it;
}
bool nodo::operator==(const nodo& n)const   // operator== per singolo nodo
{
    return info==n.info;
}
bool nodo::operator!=(const nodo& n)const
{
    return n.info!=info;
}
nodo& nodo::operator =(const nodo& n)    // operator= per copiare l'info di n
{
    if(this != &n)
    {
        info=n.info;
    }
    return *this;
}
nodo& nodo::operator =(double n)    // operator= per copiare l'info di n
{
    info=n;
    return *this;
}
bool nodo::equals(nodo* n)const    // operator== profondo
{
    if(info!=n->info || numeroFigli()!=n->numeroFigli())
        return false;
    list<nodo*>::const_iterator it=nodi.begin();
    list<nodo*>::const_iterator it1=n->nodi.begin();
    bool ris=true;
    for(;it!=nodi.end() && ris;it++,it1++)
        ris=(*it)->equals(*it1);
    return ris;
}
nodo* nodo::copyHelp(nodo* n, nodo* p)
{
    nodo* temp=new nodo();
    temp->info=n->info;

    temp->padre=p;
    list<nodo*>::const_iterator it=n->nodi.begin();
    for(;it!=n->nodi.end();it++)
        if(*it)
            temp->nodi.push_back((*it)->copy());
    return temp;
}
nodo* nodo::copy()    // ritorna una copia del sottoalbero radicato in this
{
    return copyHelp(this,0);
}
bool nodo::contenuto(double i)const       // controlla se i è contenuto nel sottoalbero radicato in this
{
    bool trovato=false;
    if(this!=NULL)
    {
        if(info==i)
            return true;
        if(nodi.size()>0)
        {
            list<nodo*>::const_iterator it=nodi.begin();
            for(;it!=nodi.end() && !trovato;it++)
                trovato=(*it)->contenuto(i);
        }
    }
    return trovato;
}
bool nodo::contenuto(nodo* n)const       // controlla se n è contenuto nel sottoalbero radicato in this
{
    bool trovato=false;
    if(this)
    {
        if(this==n)
            return true;
        if(nodi.size()>0)
        {
            list<nodo*>::const_iterator it=nodi.begin();
            for(;it!=nodi.end() && !trovato;it++)
                trovato=(*it)->contenuto(n);
        }
    }
    return trovato;
}
int nodo::profondita()const   // ritorna la profondità dell'albero
{
    int temp=0,temp1=0;
    list<nodo*>::const_iterator it=nodi.begin();
    for(;it!=nodi.end();it++)
    {
        if(*it)
            temp1=(*it)->profondita();
        if(temp1>temp)
            temp=temp1;
    }
    return temp+1;
}
void nodo::aggiungiNodo(nodo* n)
{
    nodi.push_back(n);
}
bool nodo::isFoglia()const
{
    if(nodi.size()==0)
        return true;
    list<nodo*>::const_iterator it=nodi.begin();
    for(;it!=nodi.end();it++)
        if(nodi.size() && *it!=NULL)
            return false;
    return true;
}
int nodo::nodiLiv(int liv)const
{
    if(!liv)
        return 1;
    if(liv>profondita() || nodi.size()==0)
        return 0 ;
    int sum=0;
    list<nodo*>::const_iterator it=nodi.begin();
    for(;it!=nodi.end();it++)
        if(*it)
            sum+=(*it)->nodiLiv(liv-1);
    return sum;
}

std::string nodo::indirizzo() const
{

    const void * address = static_cast<const void*>(this);
    std::ostringstream ss;
    ss << address;
    std::string s=ss.str();
    return s;
}
void nodo::distruggi()
{
    if(padre)
    {
        list<nodo*>::iterator it1=padre->nodi.begin();
        while (*it1!=this)
            it1++;
        *it1=NULL;
        padre=NULL;
    }
    delete this;
}

nodo* nodo::trovaNodo(QString n)
{
    if(indirizzo()==n.toStdString())
        return this;
    nodo* ris=NULL;
    for(list<nodo*>::const_iterator it=nodi.begin();it!=nodi.end() && ris==NULL;it++)
        if(*it)
            ris=(*it)->trovaNodo(n);
    return ris;
}
void nodo::stacca()
{
    for(list<nodo*>::iterator it=padre->nodi.begin();it!=padre->nodi.end();it++)
        if(this==*it)
        {
            *it=NULL;
            break;
        }

    padre=NULL;
    for(list<nodo*>::iterator it=nodi.begin();it!=nodi.end();it++)
        *it=NULL;
}

File: java/nodo.java

import java.util.ArrayList;
import java.util.List;

public class nodo
{
    private double info;
    private nodo padre;
    private List<nodo> nodi=new ArrayList<nodo>();;
    
    public nodo()
    {
        padre=null;
        info=0;
    }
    public nodo(nodo n)
    {
        info=n.info;
        padre=n.padre;
        /*for(int i=0;i<nodi.size();i++)
        {
            nodi.add(n.nodi.get(i).copy());
            nodi.get(i).setPadre(this);
        }*/
    }
    public nodo(int num)
    {
        padre=null;
        info=0;
        for(int i=0;i<num;i++)
        {
            nodi.add(new nodo());
            nodi.get(i).setPadre(this);
        }
    }
    private nodo copyHelp(nodo n,nodo p)
    {
        nodo temp=new nodo();
        temp.info=n.info;
        temp.padre=p;
        
        for(int i=0;i<n.nodi.size();i++)
        {
            if(n.nodi.get(i)!=null)
            {
                nodi.add(copyHelp(n.nodi.get(i),temp));
                nodi.get(i).setPadre(this);
            }
            else
                nodi.add(null);
        }
        return temp;
    }
    public nodo copy()
    {
        return copyHelp(this,null);
    }
    public nodo(double i,nodo p,List<nodo> n)
    {
        info=i;
        padre=p;
        nodi=n;
    }
    public nodo(double i,nodo p)
    {
        info=i;
        padre=p;
    }
    public int numeroFigli()
    {
        return nodi.size();
    }
    public nodo getNodo(int i) throws exc
    {
        if(i>nodi.size())
            throw new exc("Nodo non esistente.");
        return nodi.get(i);
    }
    public nodo getPadre()
    {
        return padre;
    }
    public double getInfo()
    {
        return info;
    }
    public void setInfo(double i)
    {
        info=i;
    }
    public void setPadre(nodo p)
    {
        padre=p;
    }
    public void setFiglio(int i,nodo n)
    {
        if(i>nodi.size())
            throw new exc("Nodo non esistente.");
        nodi.set(i,n);
    }
    public boolean equals(nodo n)
    {
        if(info!=n.info || nodi.size()!=n.nodi.size())
            return false;
        boolean ris=true;
        for(int i=0;i<nodi.size() && ris;i++)
            if((nodi.get(i)== null && n.nodi.get(i)!=null) || (nodi.get(i)!= null && n.nodi.get(i)==null))
                return false;
            else if(nodi.get(i)!= null && n.nodi.get(i)!=null)
                ris=nodi.get(i).equals(n.nodi.get(i));
        return ris;
    }
    public boolean contenuto(double i)
    {
        if(i==info)
            return true;
        boolean ris=false;
        for(int it=0;it<nodi.size() && !ris;it++)
            if(nodi.get(it)!=null)
                ris=nodi.get(it).contenuto(i);
        return ris;
    }
    public boolean contenuto(nodo n)
    {
        if(this==n)
            return true;
        boolean ris=false;
        for(int it=0;it<nodi.size() && !ris;it++)
            if(nodi.get(it)!=null)
                ris=nodi.get(it).contenuto(n);
        return ris;
    }
    public int profondita()
    {
        int t1=0,t2=0;
        for(int i=0;i<nodi.size();i++)
        {
            if(nodi.get(i)!=null)
                t1=nodi.get(i).profondita();
            if(t1>t2)
                t2=t1;
        }
        return t2+1;
    }
    public void aggiungiNodo(nodo n)
    {
        nodi.add(n);
        //n.padre=this;
    }
    public boolean isFoglia()
    {
        if(nodi.size()==0)
            return true;
        for(int i=0;i<nodi.size();i++)
            if(nodi.get(i)!=null)
                return false;
        return true;
    }
    public int nodiLiv(int liv)
    {
        if(liv==0)
            return 1;
        if(liv>profondita() || nodi.size()==0)
            return 0 ;
        int sum=0;
        for(int i=0;i<nodi.size();i++)
            if(nodi.get(i)!=null)
                sum+=nodi.get(i).nodiLiv(liv-1);
        return sum;
    }
    public void distruggi()
    {
        for(int i=0;i<nodi.size();i++)
            if(nodi.get(i)!=null)
                nodi.get(i).distruggi();
        int it=0;
        if(padre!=null)
        {
            while (padre.nodi.get(it)!=this)
                it++;
            padre.nodi.set(it,null);
            padre=null;
        }
    }
    public void stacca()
    {
        int i=0;
        for(;i<padre.nodi.size();i++)
            if(this==padre.nodi.get(i))
            {
                padre.nodi.set(i,null);
                break;
            }
        padre = null;
        i=0;
        for(;i<nodi.size();i++)
        {
            if(nodi.get(i)!=null)
            {
                nodi.get(i).setPadre(null);
                nodi.set(i,null);
            }
        }
    }
    private void stampa1()
    {
        if(padre!=null && padre.nodi.get(padre.nodi.size()-1)==this)
            System.out.print("(" + info );
        else
            System.out.print("(" + info );
        int i=0;
        for(;i<nodi.size();i++)
        {
            if(nodi.get(i)!=null)
                nodi.get(i).stampa1();
            else
                if(i==nodi.size()-1)
                    System.out.print("()");
                else
                    System.out.print("(),");
        }
        System.out.print(")");
    }
    public void stampa()
    {
        System.out.println();
        stampa1();
        
    }
}
